File: app/channels/lid_resolver.py
"""LID (List ID) resolution — 7 strategies, scoped per WhatsApp account.

Resolves WhatsApp LID JIDs (e.g., 12345@lid) to real phone numbers.
All caching and DB lookups are scoped by whatsapp_account_id to prevent
cross-tenant data leaks.

Strategies (in order):
1. Memory cache (scoped by account_id + lid_jid)
2. DB mapping table
3. Contacts API profilePicUrl match
4. Contacts API pushName match (unique only)
5. Evolution DB Contact table (profilePicUrl)
6. Evolution DB Contact table (pushName)
7. Message timestamp correlation
"""

import logging
from app.channels import whatsapp
from app.db import lid as lid_db

log = logging.getLogger('channels.lid_resolver')

# In-memory cache: (whatsapp_account_id, lid_jid) -> phone
_cache = {}


def _same_profile_pic(url1, url2):
    if not url1 or not url2:
        return False
    return url1.split('?')[0] == url2.split('?')[0]


def _save(account_id, lid_jid, phone, instance_name, push_name, source):
    """Save resolved mapping to cache + DB. Respects source priority."""
    result = lid_db.save_mapping(account_id, lid_jid, phone, source, push_name)
    if result is None:
        # Blocked by priority — don't update cache either
        log.info(f'LID save skipped (low priority {source}): {lid_jid} -/-> {phone}')
        return False
    _cache[(str(account_id), lid_jid)] = phone
    log.info(f'LID resolved via {source}: {lid_jid} -> {phone}')
    return True


def resolve(whatsapp_account_id, instance_name, lid_jid):
    """Resolve a LID JID to a real phone number using 7 strategies.

    Returns phone string or None if unresolvable.
    """
    account_id = str(whatsapp_account_id)
    cache_key = (account_id, lid_jid)

    # Strategy 1: Memory cache (validated against DB to catch manual corrections)
    if cache_key in _cache:
        cached_phone = _cache[cache_key]
        try:
            db_phone = lid_db.get_phone(account_id, lid_jid)
            if db_phone and db_phone != cached_phone:
                log.warning(
                    f'Cache STALE for {lid_jid}: cache={cached_phone}, DB={db_phone}. Using DB.'
                )
                _cache[cache_key] = db_phone
                return db_phone
        except Exception:
            pass
        return cached_phone

    # Strategy 2: DB mapping
    try:
        db_phone = lid_db.get_phone(account_id, lid_jid)
        if db_phone:
            _cache[cache_key] = db_phone
            log.info(f'LID resolved via DB: {lid_jid} -> {db_phone}')
            return db_phone
    except Exception:
        pass

    # Strategies 3+4: Contacts API
    push_name = ''
    try:
        contacts = whatsapp.fetch_all_contacts(instance_name)
        if contacts:
            lid_contact = None
            for c in contacts:
                if c.get('remoteJid') == lid_jid:
                    lid_contact = c
                    break

            if lid_contact:
                pic_url = lid_contact.get('profilePicUrl')
                push_name = lid_contact.get('pushName', '')

                # Strategy 3: profilePicUrl match
                if pic_url:
                    for c in contacts:
                        rjid = c.get('remoteJid', '')
                        if (rjid.endswith('@s.whatsapp.net')
                                and _same_profile_pic(c.get('profilePicUrl'), pic_url)):
                            phone = rjid.split('@')[0]
                            _save(account_id, lid_jid, phone, instance_name, push_name, 'profilePic API')
                            return phone

                # Strategy 4: pushName (unique match only)
                if push_name:
                    candidates = [
                        c for c in contacts
                        if c.get('remoteJid', '').endswith('@s.whatsapp.net')
                        and c.get('pushName') == push_name
                    ]
                    if len(candidates) == 1:
                        phone = candidates[0]['remoteJid'].split('@')[0]
                        _save(account_id, lid_jid, phone, instance_name, push_name, 'pushName API')
                        return phone
    except Exception:
        pass

    # Strategies 5+6: Evolution DB Contact table
    try:
        phone = lid_db.resolve_via_evolution_db_contact(lid_jid)
        if phone:
            _save(account_id, lid_jid, phone, instance_name, push_name, 'Evolution DB Contact')
            return phone
    except Exception:
        pass

    # Strategy 7: Message timestamp correlation
    try:
        phone = lid_db.resolve_via_message_correlation(lid_jid)
        if phone:
            _save(account_id, lid_jid, phone, instance_name, push_name, 'msg correlation')
            return phone
    except Exception:
        pass

    log.warning(f'LID unresolved (7 strategies): {lid_jid} (push={push_name})')
    return None


def learn_from_sent_message(whatsapp_account_id, instance_name, data):
    """Learn LID-phone mapping from an outgoing message."""
    account_id = str(whatsapp_account_id)
    try:
        remote_jid = data.get('key', {}).get('remoteJid', '')
        if '@s.whatsapp.net' not in remote_jid:
            return
        phone = remote_jid.split('@')[0]
        push_name = data.get('pushName', '')

        contacts = whatsapp.fetch_all_contacts(instance_name)
        if not contacts:
            return

        sent_contact = None
        for c in contacts:
            if c.get('remoteJid') == remote_jid:
                sent_contact = c
                break
        if not sent_contact:
            return

        pic = sent_contact.get('profilePicUrl')
        pn = sent_contact.get('pushName', '') or push_name

        if pic:
            for c in contacts:
                rjid = c.get('remoteJid', '')
                if '@lid' in rjid and _same_profile_pic(c.get('profilePicUrl'), pic):
                    _save(account_id, rjid, phone, instance_name, pn or c.get('pushName', ''), 'sent profilePic')
                    return

        if pn:
            lid_candidates = [
                c for c in contacts
                if '@lid' in c.get('remoteJid', '') and c.get('pushName') == pn
            ]
            if len(lid_candidates) == 1:
                rjid = lid_candidates[0]['remoteJid']
                if (account_id, rjid) not in _cache:
                    _save(account_id, rjid, phone, instance_name, pn, 'sent pushName')
    except Exception:
        pass


def learn_from_contacts_event(whatsapp_account_id, instance_name, data):
    """Learn LID-phone mapping from contacts.upsert/update events."""
    account_id = str(whatsapp_account_id)
    try:
        contacts = data if isinstance(data, list) else [data]
        for contact in contacts:
            if not isinstance(contact, dict):
                continue
            contact_id = contact.get('id', '') or contact.get('remoteJid', '')
            lid = contact.get('lid', '')
            push_name = (
                contact.get('name')
                or contact.get('notify', '')
                or contact.get('pushName', '')
            )
            if (contact_id and '@s.whatsapp.net' in contact_id
                    and lid and '@lid' in lid):
                phone = contact_id.split('@')[0]
                _save(account_id, lid, phone, instance_name, push_name, 'contacts event')
                return lid, phone
    except Exception:
        pass
    return None, None

File: agent-node/src/services/lid-resolver.js
/**
 * LID (Linked ID) Resolver
 * Resolves WhatsApp LID JIDs (e.g., 12345@lid) to real phone numbers
 * for Instagram/Facebook Messenger integration
 */

const { getDatabase } = require('../db/sqlite');
const logger = require('../utils/logger');
const axios = require('axios');
const config = require('../config');
const { Pool } = require('pg');

// In-memory cache: lid_jid -> phone
const cache = new Map();

// PostgreSQL connection for fallback lookup
const pgPool = new Pool({
  connectionString: process.env.DATABASE_URL,
  max: 2,
  idleTimeoutMillis: 30000,
});

class LIDResolver {
  constructor() {
    this.evolutionUrl = config.evolution.url;
    this.apiKey = config.evolution.apiKey;
  }

  /**
   * Check if a JID is a LID (Linked ID)
   */
  isLID(jid) {
    return jid && jid.includes('@lid');
  }

  /**
   * Get the correct JID for sending messages
   * If it's a LID, try to resolve it; otherwise return as-is
   */
  async getDestinationJid(instanceName, remoteJid) {
    if (!this.isLID(remoteJid)) {
      // Regular WhatsApp number, use as-is
      return remoteJid;
    }

    // Try to resolve LID to phone number
    const phone = await this.resolve(instanceName, remoteJid);
    if (phone) {
      return `${phone}@s.whatsapp.net`;
    }

    // If we can't resolve, return original (Evolution might handle it)
    logger.warn(`Could not resolve LID: ${remoteJid}, using original`);
    return remoteJid;
  }

  /**
   * Resolve a LID JID to a real phone number
   */
  async resolve(instanceName, lidJid) {
    // Check cache first
    if (cache.has(lidJid)) {
      return cache.get(lidJid);
    }

    // Check database
    const db = getDatabase();
    try {
      const mapping = db.prepare(`
        SELECT phone FROM lid_mappings WHERE lid_jid = ?
      `).get(lidJid);

      if (mapping?.phone) {
        cache.set(lidJid, mapping.phone);
        logger.debug(`LID resolved from DB: ${lidJid} -> ${mapping.phone}`);
        return mapping.phone;
      }
    } catch (e) {
      // Table might not exist, create it
      this.ensureTable();
    }

    // Try PostgreSQL as fallback (main database)
    try {
      const pgResult = await pgPool.query(
        'SELECT phone, push_name FROM lid_mappings WHERE lid_jid = $1',
        [lidJid]
      );
      if (pgResult.rows.length > 0) {
        const { phone, push_name } = pgResult.rows[0];
        // Sync to SQLite for faster future lookups
        this.saveMapping(lidJid, phone, push_name, 'postgres_sync');
        logger.info(`[OK] LID resolved from PostgreSQL: ${lidJid} -> ${phone}`);
        return phone;
      }
    } catch (pgError) {
      logger.debug('PostgreSQL LID lookup failed:', pgError.message);
    }

    // Try to resolve via contacts API
    try {
      const contacts = await this.fetchContacts(instanceName);
      if (contacts && contacts.length > 0) {
        const lidContact = contacts.find(c => c.remoteJid === lidJid);

        if (lidContact) {
          const picUrl = lidContact.profilePicUrl;
          const pushName = lidContact.pushName;

          // Strategy 1: Match by profile picture URL
          if (picUrl) {
            for (const c of contacts) {
              if (c.remoteJid?.endsWith('@s.whatsapp.net') &&
                  this.sameProfilePic(c.profilePicUrl, picUrl)) {
                const phone = c.remoteJid.split('@')[0];
                this.saveMapping(lidJid, phone, pushName, 'profilePic');
                return phone;
              }
            }
          }

          // Strategy 2: Match by unique pushName
          if (pushName) {
            const candidates = contacts.filter(c =>
              c.remoteJid?.endsWith('@s.whatsapp.net') &&
              c.pushName === pushName
            );

            if (candidates.length === 1) {
              const phone = candidates[0].remoteJid.split('@')[0];
              this.saveMapping(lidJid, phone, pushName, 'pushName');
              return phone;
            }
          }
        }
      }
    } catch (error) {
      logger.error('LID resolution error:', error.message);
    }

    logger.warn(`LID unresolved: ${lidJid}`);
    return null;
  }

  /**
   * Compare profile picture URLs (ignoring query params)
   */
  sameProfilePic(url1, url2) {
    if (!url1 || !url2) return false;
    return url1.split('?')[0] === url2.split('?')[0];
  }

  /**
   * Fetch contacts from Evolution API
   */
  async fetchContacts(instanceName) {
    try {
      const response = await axios.post(
        `${this.evolutionUrl}/chat/findContacts/${instanceName}`,
        {},
        {
          headers: {
            'apikey': this.apiKey,
            'Content-Type': 'application/json',
          },
          timeout: 10000,
        }
      );
      return response.data || [];
    } catch (error) {
      logger.error('Failed to fetch contacts:', error.message);
      return [];
    }
  }

  /**
   * Save LID -> phone mapping to database (SQLite + PostgreSQL)
   */
  saveMapping(lidJid, phone, pushName, source) {
    const db = getDatabase();
    this.ensureTable();

    // Save to SQLite (local cache)
    try {
      db.prepare(`
        INSERT OR REPLACE INTO lid_mappings (lid_jid, phone, push_name, source, updated_at)
        VALUES (?, ?, ?, ?, datetime('now'))
      `).run(lidJid, phone, pushName || '', source);

      cache.set(lidJid, phone);
      logger.info(`[OK] LID resolved via ${source}: ${lidJid} -> ${phone}`);
    } catch (error) {
      logger.error('Failed to save LID mapping to SQLite:', error.message);
    }

    // Also save to PostgreSQL (main database) for persistence
    if (source !== 'postgres_sync') {
      this.saveMappingToPostgres(lidJid, phone, pushName, source).catch(e =>
        logger.debug('PostgreSQL LID save failed:', e.message)
      );
    }
  }

  /**
   * Save mapping to PostgreSQL
   */
  async saveMappingToPostgres(lidJid, phone, pushName, source) {
    try {
      // Get the default whatsapp_account_id
      const waResult = await pgPool.query(
        'SELECT id FROM whatsapp_accounts LIMIT 1'
      );
      if (waResult.rows.length === 0) return;

      const waId = waResult.rows[0].id;

      await pgPool.query(`
        INSERT INTO lid_mappings (whatsapp_account_id, lid_jid, phone, push_name, resolved_via, created_at)
        VALUES ($1, $2, $3, $4, $5, NOW())
        ON CONFLICT (lid_jid, whatsapp_account_id) DO UPDATE SET
          phone = EXCLUDED.phone,
          push_name = EXCLUDED.push_name,
          resolved_via = EXCLUDED.resolved_via
      `, [waId, lidJid, phone, pushName || '', source]);

      logger.debug(`[OK] LID saved to PostgreSQL: ${lidJid} -> ${phone}`);
    } catch (error) {
      throw error;
    }
  }

  /**
   * Learn from incoming message (contacts.upsert event)
   */
  learnFromContact(contactData) {
    try {
      const contactId = contactData.id || contactData.remoteJid || '';
      const lid = contactData.lid || '';

      if (contactId.includes('@s.whatsapp.net') && lid.includes('@lid')) {
        const phone = contactId.split('@')[0];
        const pushName = contactData.name || contactData.notify || contactData.pushName || '';
        this.saveMapping(lid, phone, pushName, 'contacts_event');
        return { lid, phone };
      }
    } catch (error) {
      logger.error('Learn from contact error:', error.message);
    }
    return null;
  }

  /**
   * Ensure lid_mappings table exists
   */
  ensureTable() {
    const db = getDatabase();
    db.exec(`
      CREATE TABLE IF NOT EXISTS lid_mappings (
        lid_jid TEXT PRIMARY KEY,
        phone TEXT NOT NULL,
        push_name TEXT,
        source TEXT,
        updated_at DATETIME DEFAULT CURRENT_TIMESTAMP
      )
    `);
  }
}

module.exports = new LIDResolver();
